### DISCUSSION TITLE
Why is _Binary not an official "CamelCase" abstract type?

### DISCUSSION THREAD

**Discussion by tboddyspargo** [sqlalchemy/sqlalchemy#11977] [Usage Questions]
My question comes from a situation where I'm trying to find type correspondence between MySQL and PostgreSQL. My general strategy for doing this between different backends is:

```python
# Using the column types from the mysql table
mysql_types = [mysql.types.VARBINARY(5)]

# Get the abstract/generic types
generic_types = [t.as_generic() for t in mysql_types]

# Try to get postgres-specific types from the generic ones
pg_types = [t.dialect_impl(postgresql.dialect() for t in generic_types)]

# Create a table object
table = sa.Table('my_table', sa.MetaData(), *[sa.Column(f"col{i}", t) for it in pg_types.enumerate()])

# Create the table itself
table.create(engine)
```

While this strategy has served me well, is fails with `VARBINARY` because `as_generic` returns itself, instead of an abstract class. The postgresql TypeConverter doesn't have any `visit_VARBINARY` method defined, so it compiles a CreateColumn statement to `col0 VARBINARY` which results in a database error due to an unrecognized type.

In contrast, this strategy works as I expect with `mysql.types.BLOB` because it _does_ inherit from an abstract type (`LargeBinary`) that's visitable and that postgresql has a `visit_large_binary` method defined for to correctly render the column statement as `col0 BYTEA`.

Since many dialects implement subclasses of `sqltypes._Binary`, shouldn't it be a valid return type for any of those subclasses `as_generic` methods? Furthermore, shouldn't dialects (like postgresql) implement `visit__binary` methods to render their appropriate corresponding datatypes and add `dialect.colspecs` entries to map it to their implementation-specific type classes?

Thanks in advance for your help on this!

**Comment by zzzeek** [comment]
So the problem can be stated more simply as this:

```py
>>> mysql.VARBINARY().as_generic()
VARBINARY()

```

an as_generic()  that doesnt actually return a generic is not helpful.

`_Binary` is a private type and doesnt really have any meaningful difference between itself and `LargeBinary`.    

I think the best behavior would be this:

```py
>>> mysql.VARBINARY().as_generic()
LargeBinary()

```

then you could work with that.    How about we do that?

**Reply by tboddyspargo** [reply]
Great summary! Yes, that would work wonderfully. 

However, are you thinking that the solution will require all dialects to update the parent class of their respective binary types? Or is there a more transparent way you were thinking of accomplishing this?

**Reply by zzzeek** [reply]
I would hardcode something onto `_Binary` that makes sure it upgrades to `LargeBinary` when the methods in this area are called

**Reply by tboddyspargo** [reply]
Sounds great! Thanks, @zzzeek !

**Reply by zzzeek** [reply]
see #11978
